File: common/src/main/java/generations/gg/generations/core/generationscore/world/item/legends/ElementalPostBattleUpdateItemImplImpl.java
package generations.gg.generations.core.generationscore.world.item.legends;

import com.cobblemon.mod.common.api.types.ElementalType;
import generations.gg.generations.core.generationscore.config.SpeciesKey;
import generations.gg.generations.core.generationscore.world.entity.block.PokemonUtil;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;

public class ElementalPostBattleUpdateItemImplImpl extends ElementalPostBattleUpdateItemImpl {
    public static final String DEFAULT_LANG_KEY = "generations_core.orb.amountfull";

    private final String lang;
    private final SpeciesKey key;

    public ElementalPostBattleUpdateItemImplImpl(Properties properties, SpeciesKey key, ElementalType... types) {
        this(properties, DEFAULT_LANG_KEY, key, types);
    }


    public ElementalPostBattleUpdateItemImplImpl(Properties properties, String lang, SpeciesKey key, ElementalType... types) {
        super(properties, types);
        this.lang = lang;
        this.key = key;
    }

    @Override
    public @NotNull InteractionResultHolder<ItemStack> use(Level level, Player player, @NotNull InteractionHand usedHand) {
        var stack = player.getItemInHand(usedHand);

        if (!level.isClientSide()) {
            int damage = stack.getDamageValue();

            if (damage >= getMaxDamage()) {
                stack.shrink(1);
                PokemonUtil.spawn(key.createProperties(70), level, player.getOnPos());
                postSpawn(level, player, usedHand);
            } else {
                player.displayClientMessage(Component.translatable(lang, getMaxDamage() - damage), true);
            }

            return InteractionResultHolder.success(stack);
        }

        return InteractionResultHolder.pass(stack);
    }

    protected void postSpawn(Level level, Player player, InteractionHand usedHand) {
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/item/legends/ElementalPostBattleUpdateItemImplImpl.kt
package generations.gg.generations.core.generationscore.common.world.item.legends

import com.cobblemon.mod.common.api.types.ElementalType
import generations.gg.generations.core.generationscore.common.config.SpeciesKey
import generations.gg.generations.core.generationscore.common.world.entity.block.PokemonUtil
import net.minecraft.core.Holder
import net.minecraft.network.chat.Component
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResultHolder
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level

open class ElementalPostBattleUpdateItemImplImpl(
    properties: Properties,
    private val lang: String = DEFAULT_LANG_KEY,
    private val key: SpeciesKey,
    types: List<ElementalType> = listOf(),
    private val itemToGiveUponSpawn : Holder<Item>? = null,
    private val predicate: (Player) -> Boolean = { true }
    ) : ElementalPostBattleUpdateItemImpl(properties, types) {

    override fun use(level: Level, player: Player, usedHand: InteractionHand): InteractionResultHolder<ItemStack> {
        val stack = player.getItemInHand(usedHand)

        if (!level.isClientSide() && predicate.invoke(player)) {

            val damage = stack.damageValue

            if (damage >= stack.maxDamage) {
                PokemonUtil.spawn(key.createProperties(70), level, player.onPos)
                stack.consume(1, player)
                postSpawn(player)
            } else {
                player.displayClientMessage(Component.translatable(lang, stack.maxDamage - damage), true)
            }

            return InteractionResultHolder.success(stack)
        }

        return InteractionResultHolder.pass(stack)
    }

    protected fun postSpawn(player: Player) {
        itemToGiveUponSpawn?.value()?.defaultInstance?.also(player::addItem)
    }

    companion object {
        const val DEFAULT_LANG_KEY: String = "generations_core.orb.amountfull"
    }
}
